### PR TITLE
CLI for threshold plugin

### PR DESCRIPTION
**Pull request by bayliffe**
Issue #98 

A CLI to accept a single or multiple thresholds to apply to a cube. When multiple thresholds are provided it produces a concatenated cube with a threshold dimension coordinate.

Acceptance tests to be added following first review.

### CODE REVIEW

File: bin/improver-threshold [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        "threshold.")
+    parser.add_argument("input_filepath", metavar="INPUT_FILE",
+                        help="A path to an input NetCDF file to be processed")
+    parser.add_argument("output_filepath", metavar="OUTPUT_FILE",
+                        help="The output path for the processed NetCDF")
+    parser.add_argument("threshold_values", metavar="THRESHOLD_VALUES",
+                        nargs="+", type=float,
+                        help="Threshold value or values about which to "
+                        "calculate the truth values; e.g. 270 300")
+    parser.add_argument("--threshold_units", metavar="THRESHOLD_UNTIS",
```
Comment (fionaRust) [inline bin/improver-threshold]: `THRESHOLD_UNITS` not  `THRESHOLD_UNTIS`

File: bin/improver-threshold [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+from improver.threshold import BasicThreshold
+from improver.utilities.cube_manipulation import concatenate_cubes
+from improver.spotdata.extract_data import ExtractData
+
+
+def main():
+    """Load in arguments and get going."""
+    parser = argparse.ArgumentParser(
+        description="Calculate the threshold truth value of cube data "
```
Comment (fionaRust) [inline bin/improver-threshold]: Maybe we should put the default behaviour of above threshold in the description?

File: bin/improver-threshold [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+                        "fuzzy. Data which fail a test against the hard "
+                        "threshold value may return a fractional truth value "
+                        "if they fall within this fuzzy factor region.")
+
+    args = parser.parse_args()
+    cube = iris.load_cube(args.input_filepath)
+
+    # Allow for threshold value unit conversion.
+    if args.threshold_units is not None:
+        threshold_unit = cf_units.Unit(args.threshold_units)
```
Comment (fionaRust) [inline bin/improver-threshold]: I'm guessing we are relying on the default error to be raised if the conversion can't happen. I think this is ok.

File: bin/improver-threshold (hunk 1 above)
Comment (bayliffe) [reply bin/improver-threshold]: :white_check_mark:

File: bin/improver-threshold (hunk 3 above)
Comment (bayliffe) [reply bin/improver-threshold]: We are.

File: bin/improver-threshold [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        args.threshold_values = [threshold_unit.convert(threshold, cube.units)
+                                 for threshold in args.threshold_values]
+        print args.threshold_values
```
Comment (fionaRust) [inline bin/improver-threshold]: Need to get rid of this print statement.

Comment (bayliffe) [reply bin/improver-threshold]: :white_check_mark:

File: bin/improver-threshold [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    cube = iris.load_cube(args.input_filepath)
+
+    # Allow for threshold value unit conversion.
+    if args.threshold_units is not None:
+        threshold_unit = cf_units.Unit(args.threshold_units)
+        args.threshold_values = [threshold_unit.convert(threshold, cube.units)
+                                 for threshold in args.threshold_values]
+
+    thresholded_cubes = iris.cube.CubeList()
+    for threshold in args.threshold_values:
```
Comment (MoseleyS) [inline bin/improver-threshold]: I would expect this level of behaviour to be in the plugin, not the CLI. Does it matter where it is?

File: tests/improver-threshold/05-threshold_units.bats [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+  # Run threshold processing and check it passes.
+  run improver threshold \
+      "$IMPROVER_ACC_TEST_DIR/threshold/basic/input.nc" "$TEST_DIR/output.nc" \
+      6.85 --threshold_units celsius
+  [[ "$status" -eq 0 ]]
+
+  # Run nccmp to compare the output and kgo.
+  improver_compare_output "$TEST_DIR/output.nc" \
+      "$IMPROVER_ACC_TEST_DIR/threshold/threshold_units/kgo.nc"
```
Comment (MoseleyS) [inline tests/improver-threshold/05-threshold_units.bats]: `threshold_units/kgo.nc` is identical to `basic/kgo.nc`. Consider using the basic version here.

File: bin/improver-threshold (hunk 5 above)
Comment (fionaRust) [reply bin/improver-threshold]: We will fix this in a separate PR.

File: tests/improver-threshold/05-threshold_units.bats (hunk 6 above)
Comment (fionaRust) [reply tests/improver-threshold/05-threshold_units.bats]: We will use symlinks to deal with this.

Comment (MoseleyS) [reply tests/improver-threshold/05-threshold_units.bats]: I have changed the kgo file to a symlink and checked the CLI test still works.

**Pull request merged**